### PR TITLE
[USM] fix incompleteBuffer with 5 tuples (disable NAT) with packet timestamp and per protocol

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

This PR adding some fix for the http incompleteBuffer
 * fixing connection collision due to different protocol (http, http2)
 * avoid connection collision by matching 5 tuples (this would disable NAT support, another PR will address that as it's more complicated issue)
 * cleanup ageing requests based on packet timestamp (kernel boot time under linux, not unix epoch)
 * adding telemetry on the http joiner

### Motivation

Avoiding joining request / responses that are not related and creating high latency spike

### Additional Notes

The joiner will be able to fix this kind of situation
```
2023-07-20 18:29:08 UTC | SYS-PROBE | WARN | (pkg/network/protocols/http/statkeeper.go:68 in Process) | http incomplete : ebpfTx{Started: '42643716475756', LastSeen: '0', ResponseStatusCode: '0', Method: 'GET', Tags: '0x0', Fragment: 'frag request', Tuple: 'types.ConnectionKey{SrcIPHigh:0x0, SrcIPLow:0xb057800a, DstIPHigh:0x0, DstIPLow:0xfea9fea9, SrcPort:0x9192, DstPort:0x50}', }

2023-07-20 18:29:08 UTC | SYS-PROBE | WARN | (pkg/network/protocols/http/statkeeper.go:68 in Process) | http incomplete : ebpfTx{Started: '0', LastSeen: '42643717221365', ResponseStatusCode: '200', Method: 'UNKNOWN', Tags: '0x0', Fragment: 'frag response', Tuple: 'types.ConnectionKey{SrcIPHigh:0x0, SrcIPLow:0xb057800a, DstIPHigh:0x0, DstIPLow:0xfea9fea9, SrcPort:0x9192, DstPort:0x50}', }
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/types/connection_key.go [hunk 1]
```diff
@@ -5,7 +5,10 @@
 
 package types
 
-import "github.com/DataDog/datadog-agent/pkg/process/util"
```
Comment (hmahmood) [inline pkg/network/types/connection_key.go:8]: Could you move this to `pkg/network/protocols` please? Its only being used by USM at the moment.

**Review by hmahmood — APPROVED** [review]

File: pkg/network/protocols/http/incomplete_stats.go [hunk 2]
```diff
@@ -12,11 +12,17 @@ import (
 	"time"
 
 	"github.com/DataDog/datadog-agent/pkg/network/config"
+	"github.com/DataDog/datadog-agent/pkg/network/protocols"
 	"github.com/DataDog/datadog-agent/pkg/network/types"
 )
 
 const defaultMinAge = 30 * time.Second
 
+type connectionKey struct {
```
Comment (hmahmood) [inline pkg/network/protocols/http/incomplete_stats.go:21]: Why not move the whole `ConnectionKey` struct here?

Comment (nplanel) [reply pkg/network/protocols/http/incomplete_stats.go:21]: sorry I miss read your previous comment

Comment (hmahmood) [reply pkg/network/protocols/http/incomplete_stats.go:21]: :+1:

**Comment by guyarb** [comment]
> avoid connection collision by matching 5 tuples (this would disable NAT support, another PR will address that as it's more complicated issue)
That sentence and that impact is not something we can live with. If you prefer splitting the PRs, then please open them both together so we see the whole picture

File: pkg/network/protocols/http/incomplete_stats.go [hunk 3]
```diff
@@ -101,20 +115,21 @@ func (b *incompleteBuffer) Add(tx Transaction) {
...
+		b.telemetry.joiner.requests.Add(1)
 		parts.requests = append(parts.requests, tx)
 	} else {
+		b.telemetry.joiner.responses.Add(1)
 		parts.responses = append(parts.responses, tx)
 	}
 }
 
-func (b *incompleteBuffer) Flush(now time.Time) []Transaction {
+func (b *incompleteBuffer) Flush(nowUnixNano int64) []Transaction {
```
Comment (guyarb) [inline pkg/network/protocols/http/incomplete_stats.go:126]: the pr description does not explain this change (there is a section about timestamp, but does not reflect the actual change)

File: pkg/network/protocols/http/model.go [hunk 4]
```diff
@@ -8,12 +8,14 @@
 package http
 
 import (
+	"github.com/DataDog/datadog-agent/pkg/network/protocols"
 	"github.com/DataDog/datadog-agent/pkg/network/types"
 )
 
 type Transaction interface {
-	RequestLatency() float64
+	RequestLatency() int64
```
Comment (guyarb) [inline pkg/network/protocols/http/model.go:16]: why?

File: pkg/network/protocols/http/model_linux.go [hunk 5]
```diff
@@ -10,13 +10,29 @@ package http
...
+	"github.com/DataDog/datadog-agent/pkg/network/protocols"
 	"github.com/DataDog/datadog-agent/pkg/network/types"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+// PacketNow return the timestamp in nanoseconds relative to the kernel boot time
+func PacketNow() int64 {
+	now, err := ebpf.NowNanoseconds()
+	if err != nil {
+		log.Errorf("statkeeper ebpf.NowNanoseconds() failed %s", err)
```
Comment (guyarb) [inline pkg/network/protocols/http/model_linux.go:27]: there is not relation to statkeeper

File: pkg/network/protocols/http/telemetry.go [hunk 6]
```diff
@@ -12,6 +12,14 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+type telemetryJoiner struct {
+	requests       *libtelemetry.Counter
+	responses      *libtelemetry.Counter
+	requestDropped *libtelemetry.Counter
+	requestJoined  *libtelemetry.Counter
+	agedRequest    *libtelemetry.Counter
+}
```
Comment (guyarb) [inline pkg/network/protocols/http/telemetry.go:21]: please document the struct and its fields.

**Review by guyarb — CHANGES_REQUESTED** [review]
CRed

File: pkg/network/protocols/http/incomplete_stats.go (hunk 3 above)
Comment (nplanel) [reply pkg/network/protocols/http/incomplete_stats.go:126]: I will update the description

**Review by urseberry — APPROVED** [review]

**Comment by nplanel** [comment]
I closing this PR as it's not fixing or improve enough the main issue
incomplete buffer algorithm doesn't associate the correct request and response
an RFC will follow with the finding and possible solutions

**Pull request closed without merging**